### PR TITLE
VM Cloning design proposal

### PR DESCRIPTION
**Pull request by iholder-redhat**
A design proposal for VM cloning.

* [Design Proposal in human readable format](https://github.com/kubevirt/community/blob/98d2aeb59ff8f70a62daad435641ffb7f485f770/design-proposals/VM-cloning.md)

### CODE REVIEW

**Comment by iholder-redhat** [comment]
/cc @ormergi @EdDev @awels @alicefr @AlonaKaplan @davidvossel @vladikr @ronensdeor

Feedback is greatly appreciated.

File: design-proposals/VM-cloning.md
Comment (alicefr) [inline design-proposals/VM-cloning.md:84]: This implies that CDI needs to be installed in order to use the feature.

File: design-proposals/VM-cloning.md
Comment (alicefr) [inline design-proposals/VM-cloning.md:90]: What about [host disks](https://kubevirt.io/user-guide/virtual_machines/disks_and_volumes/#hostdisk)? They are not copied automatically

File: design-proposals/VM-cloning.md
Comment (alicefr) [inline design-proposals/VM-cloning.md:100]: I think certain annotations are required especially those not set by the user or kubevirt directly but used by other tools

File: design-proposals/VM-cloning.md
Comment (awels) [reply design-proposals/VM-cloning.md:90]: Host disks are problematic in various ways, and really should be retired.

File: design-proposals/VM-cloning.md
Comment (awels) [reply design-proposals/VM-cloning.md:84]: I don't think this is a problem, unless some other mechanism of cloning volumes is better. We can think of CSI cloning (which is seamlessly integrated with CDI already) as an alternative, but that won't help non csi storage providers.

Comment (iholder-redhat) [reply design-proposals/VM-cloning.md:84]: Maybe I should add this into the design proposal as a limitation / requirement in order for VM cloning to work. WDYT?

File: design-proposals/VM-cloning.md
Comment (iholder-redhat) [reply design-proposals/VM-cloning.md:90]: Same here - I think it's best adding to the doc that host disks wouldn't be supported for this.

BTW - since host disks should be retired - maybe we can deprecate it all-together (of course over time etc)?

Comment (awels) [reply design-proposals/VM-cloning.md:90]: Well its my opinion they should be retired. We should start a discussion, but definitely note that hostdisks would not be supported for cloning.

**Review by davidvossel — COMMENTED** [review]
@mhenriks

Snapshotting and cloning have always kind of seemed to be in the same realm. Is there any technical reason why we shouldn't think of a clone as a VM snapshot that's being restored as a brand new VM?

From there the workflow would be something like
1. Start a VM
2. create a snapshot of VM
3. Create as many clones as you want from that snapshot through something like a more advanced restore action.

**Comment by mhenriks** [comment]
> @mhenriks
> 
> Snapshotting and cloning have always kind of seemed to be in the same realm. Is there any technical reason why we shouldn't think of a clone as a VM snapshot that's being restored as a brand new VM?
> 
> From there the workflow would be something like
> 
> 1. Start a VM
> 2. create a snapshot of VM
> 3. Create as many clones as you want from that snapshot through something like a more advanced restore action.

Good question!  I think it would be great to be able to "restore" a snapshot to a new VM.  But I don't think it should be required for a VM clone to be based on a snapshot.  Cloning a VM directly will allow supporting local storage/non snapshot-able storage.  It is also a simpler experience for the user.  

But we should consider how clone/restore from VM Snapshot may be supported:

1.  `VirtualMachineClone` has a source field that may refer to a `VirtualMachineSnapshot`
2. `VirtualMachineRestore` allows target to be new VM

I'm inclined to prefer 2.  But there will be a lot of shared functionality between clone/restore like mutating existing vm specs for a new vm.

**Review by mhenriks — COMMENTED** [review]
Looks good!  Nice this initiative moving forward!

File: design-proposals/VM-cloning.md
Comment (iholder-redhat) [reply design-proposals/VM-cloning.md:60]: good idea, thanks!

File: design-proposals/VM-cloning.md
Comment (iholder-redhat) [reply design-proposals/VM-cloning.md]: > Should they be owned by the new VM?
Is there a reason for them not to be?

File: design-proposals/VM-cloning.md
Comment (rthallisey) [reply design-proposals/VM-cloning.md:89]: The CSI layer supports volume cloning.  So shouldn't the expectation be that for a storage clone to succeed, the admin must be using a plugin that supports volume cloning?

File: design-proposals/VM-cloning.md
Comment (mhenriks) [reply design-proposals/VM-cloning.md]: I guess it's possible that someone doesn't want to bind the lifecycle of their data to the VM.  That's probably the main reason why someone wouldn't use DataVolumeTemplates in the first place?

File: design-proposals/VM-cloning.md
Comment (EdDev) [inline design-proposals/VM-cloning.md:43]: You mean a `CR`?

File: design-proposals/VM-cloning.md
Comment (EdDev) [inline design-proposals/VM-cloning.md:49]: What is the default?
In general, this sounds risky to me. Some future annotation may not consider this cloning to exist and cause issues.

File: design-proposals/VM-cloning.md
Comment (EdDev) [inline design-proposals/VM-cloning.md:66]: I see only one field here, but there may be multiple interfaces to a VM.

File: design-proposals/VM-cloning.md
Comment (EdDev) [inline design-proposals/VM-cloning.md:107]: And some may not work well with cloning. The fact that we do not know, is the risky part.

File: design-proposals/VM-cloning.md
Comment (EdDev) [inline design-proposals/VM-cloning.md:45]: As having a new CRD introduces a higher level of complexity to the system, could you please share why this approach is preferred over the same logic to exist in some client/library (`virtctl`)?

BTW, unlike the migration CRD, this one does not define any status fields (i.e. success, failure and whatever other data is relevant).

File: design-proposals/VM-cloning.md
Comment (alicefr) [reply design-proposals/VM-cloning.md:89]: sorry, for the late reply. CDI is not automatically installed with KubeVirt, so if you rely on CDI then it makes it a requirement in order to use this feature.

File: design-proposals/VM-cloning.md
Comment (iholder-redhat) [reply design-proposals/VM-cloning.md:49]: The default IMO should be to copy all annotations / labels.
In the future we can exclude certain annotations that we know are problematic. Do you know of any current annotations that we should exclude?

File: design-proposals/VM-cloning.md
Comment (iholder-redhat) [reply design-proposals/VM-cloning.md:43]: Yes, will fix

File: design-proposals/VM-cloning.md
Comment (iholder-redhat) [reply design-proposals/VM-cloning.md:66]: So every interface has its own MAC address?
If so - WDYT on having a way to specify a new MAC per interface (and reference it by name)?

File: design-proposals/VM-cloning.md
Comment (iholder-redhat) [reply design-proposals/VM-cloning.md:45]: I think that using CRDs has some benefits:
- It is the "Kubernetes way" of performing such operations. It is declarative, easy to use and similar to other Kubernetes operations.
- Implementing all logic in `virtctl` implies that all of the cloning operation is performed by the client side. In general this is less robust, since the client can, for example, disconnect from the server at the middle of the cloning operation. This cannot happen if the cloning request resides in etcd and controllers watch this resource to perform the operation.
- We can have status in this CRD as well
- Some virtctl operations are actually implemented with a CRD, `virtctl migrate` for example.

> As having a new CRD introduces a higher level of complexity to the system

What is the big disadvantage here in your opinion? What are the benefits in avoiding a CRD here?

File: design-proposals/VM-cloning.md
Comment (xpivarc) [reply design-proposals/VM-cloning.md:49]: Annotation `kubevirt.io/storage-observed-api-version` & `kubevirt.io/latest-observed-api-version` comes to my mind.

File: design-proposals/VM-cloning.md
Comment (davidvossel) [reply design-proposals/VM-cloning.md:45]: the clone operation needs to be managed by the backend in a declarative way that all client side tooling can invoke. so, virtctl, kubectl, and something like the OCP UI shoudl all be able to invoke a clone operation using the same backend logic.

**Review by davidvossel — COMMENTED** [review]
/cc @mhenriks 

@iholder-redhat I think your api is headed in the right direction. I'd like to take a closer look at the implementation phases and make a conscious effort to implement this in a way that reuses the existing snapshot/restore logic. I think we can build the Clone api around the existing snapshot/restore apis. Essentially layering these APIs 

Here's a possible path forward that I think could make sense.

**Phase 1. Extend snapshot functionality to restore to new VM. (prerequisite to clone api)**

Focus entirely on making the SnapshotRestore object capable of restoring a snaptshot to a new VM. This would only involve using the existing snapshot apis and extending that functionality.

The end result here is someone could technically clone a VM by creating a snapshot and restoring to a new VM.

**Phase 2. Introduce Clone API to coordinate the workflow of a snapshot/restore to new VM**

The new Clone API would be a wrapper around the existing snapshot/restore logic and coordinate this workflow in a declarative way. This would allow someone to declare they want to Clone a VM using a single object (VirtualMachineClone) and the clone controller would coordinate creating a snapshot and restoring the snapshot to the target VM name.

The end result here is someone could post a VirtualMachineClone object targeting one source VM and get a new target VM. 

**Phase 3. Enhancements**

Today, snapshot/restore requires storage provisioners that support volume snapshots. We could techinically support offline snapshots using PVC cloning in the event that a storage provider doesn't support snapshot. By enhancing the snapshot/restore logic, the VirtualMachineClone logic would naturally get enhanced as well since the clone logic is built on snapshot/restore.

supporting Online Clones. Since we support online snapshots in some scenarios, we could extend Clone logic to be able to create clones of VMs which are online.

File: design-proposals/VM-cloning.md
Comment (EdDev) [reply design-proposals/VM-cloning.md:107]: It creates a strong dependency between new features/labels/annotations and this feature.
Such a dependency usually never ends up well. It is hard to assure it will be considered when introduced.

File: design-proposals/VM-cloning.md
Comment (rthallisey) [reply design-proposals/VM-cloning.md:89]: Then why mention CDI at all if the cloning design has nothing to do with it?

File: design-proposals/VM-cloning.md
Comment (iholder-redhat) [reply design-proposals/VM-cloning.md]: Aha, got you now. Makes sense.

I think that maybe for each one of these configs we can have a list to include / exclude stuff. It's useful also for annotations / labels / anything else. Thanks for the feedback

File: design-proposals/VM-cloning.md
Comment (iholder-redhat) [reply design-proposals/VM-cloning.md:49]: Perhaps we can solve this by having a test that expects X labels / annotations to be cloned. If we clone more than X it would fail. This way we can raise attention for added labels / annotations that are to be cloned.

In any case, I think that demanding a complete and explicit list of values to clone for every category (labels, annotations, volumes, etc) doesn't make sense - it would make cloning cumbersome and pretty useless. Furthermore, we can't demand the user to decide which of **our labels** should be cloned or not. If we're having hard time knowing it, how would the user know?

If managing our own labels / annotations is too difficult, maybe we can define several prefixes that would categorize them so it's easier to reason about their purpose. In one way or another - I think we should handle this - not the user.

File: design-proposals/VM-cloning.md
Comment (iholder-redhat) [reply design-proposals/VM-cloning.md:66]: Cool idea!

File: design-proposals/VM-cloning.md
Comment (iholder-redhat) [reply design-proposals/VM-cloning.md:45]: > Regarding being cut in the middle of the operation, that is true also when defining any other object (e.g. the VMI). But I may be missing something here.

That's right. If, for example, the creation of a VMI object would be done from `virtctl` (i.e. the client side) and would consist of say 5 steps - if the client had crashed after the 3rd step we would be stuck in an inconsistent state. This cannot happen when a VMI CR is posted to the cluster since even if some of the components crash, as soon as they are back up they would read etcd, understand that a VMI creation is stuck and what's needed to be done to complete the creation operation.

Here's a better phrasing than "kubernetes way": it is Kubernetes' solution to these kind of problems, which is an important capability.

> Isn't the cloning itself limited to just copying specific specs with some limitations?

Firstly, we're not sure what would happen in the future. Secondly, even if we just copy the spec with some limitations / modifications, we may trigger complex operations (like DV cloning). If these operations fail, in the future we may retry, fallback to different solutions, etc'.

> I think that a CRD is not cheap, it adds a great maintenance burden and should be used carefully

As I previously asked: `What is the big disadvantage here in your opinion? What are the benefits in avoiding a CRD here?`. I'm still not sure why CRD is not cheap and where is the disadvantage you're speaking about. Why do you refer it as "big guns"?

**Comment by iholder-redhat** [comment]
> /cc @mhenriks
> 
> @iholder-redhat I think your api is headed in the right direction. I'd like to take a closer look at the implementation phases and make a conscious effort to implement this in a way that reuses the existing snapshot/restore logic. I think we can build the Clone api around the existing snapshot/restore apis. Essentially layering these APIs
> 
> Here's a possible path forward that I think could make sense.
> 
> **Phase 1. Extend snapshot functionality to restore to new VM. (prerequisite to clone api)**
> 
> Focus entirely on making the SnapshotRestore object capable of restoring a snaptshot to a new VM. This would only involve using the existing snapshot apis and extending that functionality.
> 
> The end result here is someone could technically clone a VM by creating a snapshot and restoring to a new VM.
> 
> **Phase 2. Introduce Clone API to coordinate the workflow of a snapshot/restore to new VM**
> 
> The new Clone API would be a wrapper around the existing snapshot/restore logic and coordinate this workflow in a declarative way. This would allow someone to declare they want to Clone a VM using a single object (VirtualMachineClone) and the clone controller would coordinate creating a snapshot and restoring the snapshot to the target VM name.
> 
> The end result here is someone could post a VirtualMachineClone object targeting one source VM and get a new target VM.
> 
> **Phase 3. Enhancements**
> 
> Today, snapshot/restore requires storage provisioners that support volume snapshots. We could techinically support offline snapshots using PVC cloning in the event that a storage provider doesn't support snapshot. By enhancing the snapshot/restore logic, the VirtualMachineClone logic would naturally get enhanced as well since the clone logic is built on snapshot/restore.
> 
> supporting Online Clones. Since we support online snapshots in some scenarios, we could extend Clone logic to be able to create clones of VMs which are online.

Thanks @davidvossel! I really like this path forward.

I just wonder what is the motivation for using volume snapshots for cloning instead of PVC cloning. IIUC the two main reasons are:
* Supporting online VM cloning (if online snapshot is available)
* Being able to clone a VM from snapshot already captured in the past

Do you agree? Are there any reasons?

Also - PVC cloning based snapshots can benefit the snapshot API in general as IIUC some CSI drivers do not support snapshots but do support cloning.

Moving forward I will start working on snapshot API enhancements as mentioned above and keep this proposal open.

**Comment by davidvossel** [comment]
> I think @davidvossel is talking about VM snapshots, which include snapshots of all the disks (if possible).

@iholder-redhat yep, that's what I was referring to. However you're probably correct that VM snapshot does result in PVC snapshots. 


> I just wonder what is the motivation for using volume snapshots for cloning instead of PVC cloning. IIUC the two main reasons are:
>Supporting online VM cloning (if online snapshot is available)
>Being able to clone a VM from snapshot already captured in the past
>Do you agree? Are there any reasons?

My primary motivation for using VM snapshots as part of this cloning workflow is because the snapshot controllers are already doing the hard work of capturing a VM's persistent state both online and offline. If we can leverage that existing functionality to clone VMs, then we won't have to re-invent any of that logic (it was really really difficult to get that snapshot logic right, i'd hate to do something similar again)

File: design-proposals/VM-cloning.md
Comment (EdDev) [reply design-proposals/VM-cloning.md:45]: As I see it,having a CRD representing consistent & core objects like  Kubevirt cluster config, VM, VMI, etc.. makes it a great fit.
But, when you start using on a per-feature granularity, especially imperative operations, it becomes an overkill mechanism which bloats one product with too many API extensions, which are challenging to maintain.
CRD/s are more expensive to maintain because they require controllers which need to monitor them, change them, deploy, remove them, etc. It is also costly from the users side, as they need to know about them and browse through potential a large number of them in the system.

AFAIK, we already have several imperative API commands (e.g. hotplug for storage, and soon for network) which do not use CRD/s. They currently use a subresource on the virt-api entrypoint. There are also domain operations like reboot, shutdown, etc.
You could also argue that they have a flow that may be implemented using a CRD, for the same reasoning you mentioned before (e.g. being interrupted in the middle, fitting the k8s way).

I am trying to understand if there are other alternatives to extend the functionality of kubevirt without adding CRD/s. For the long run, I think it is worth exploring other options to make sure we do not overuse a k8s option, increasing kubevirt complexity beyond what is a must.
This design suggestion is presenting only one option, which makes it hard to see that other things have been explored. I think the effort to find other alternative is worth it, even if eventual a CRD will be the final choice.

To be specific, I am interested to understand why an imperative API through a subresource and potentially an annotation (in case state needs to be saved somewhere) is not enough in this case. Can't a single command on the API generate the needed VM spec and other definitions to clone the VM?

**Comment by iholder-redhat** [comment]
> > I think @davidvossel is talking about VM snapshots, which include snapshots of all the disks (if possible).
> 
> @iholder-redhat yep, that's what I was referring to. However you're probably correct that VM snapshot does result in PVC snapshots.
> 
> > I just wonder what is the motivation for using volume snapshots for cloning instead of PVC cloning. IIUC the two main reasons are:
> > Supporting online VM cloning (if online snapshot is available)
> > Being able to clone a VM from snapshot already captured in the past
> > Do you agree? Are there any reasons?
> 
> My primary motivation for using VM snapshots as part of this cloning workflow is because the snapshot controllers are already doing the hard work of capturing a VM's persistent state both online and offline. If we can leverage that existing functionality to clone VMs, then we won't have to re-invent any of that logic (it was really really difficult to get that snapshot logic right, i'd hate to do something similar again)

Makes total sense.
I've already started working on the first step forward you suggested above.

Thanks very much!

**Comment by iholder-redhat** [comment]
Proposal is updated according to @davidvossel's comment [here](https://github.com/kubevirt/community/pull/159#pullrequestreview-880329021).

@davidvossel @mhenriks can you PTAL?
Thanks

File: design-proposals/VM-cloning.md
Comment (iholder-redhat) [reply design-proposals/VM-cloning.md:46]: > To be specific, I am interested to understand why an imperative API through a subresource and potentially an annotation (in case state needs to be saved somewhere) is not enough in this case. Can't a single command on the API generate the needed VM spec and other definitions to clone the VM?

What might be problematic here IMO is:
* Complexity of arguments: if we have a large number of arguments and options, using the command line can be messy. This is especially true when providing lists of, e.g., annotations/labels/network interfaces to include/exclude. 
* We're not sure where this feature will lead us. It might be possible to clone VMIs in the future, for example. It might be limiting to have a single subresource (e.g. VM's subresource) to handle this.

In your examples, e.g. booting, restarting, migrating, etc, we don't have such problems since the operations barely have arguments and are relatively closed to expansions.

----

Another thing is that I'm not sure I fundamentally agree with the assumption that CRDs bring more complexity.

**From the performance / runtime perspective:** It's true that it's more expensive to maintain CRDs because we bring new controllers to watch and handle them, but this price certainly brings value:
* Controllers are much more reliable, robust and scalable then `virtctl`
* For developers, it's easier to maintain since controllers are Kubernetes' idiomatic way of working. `virtctl` is something different which brings more maintainability complexity from the developers' perspective.
* Controllers are optimized for distributed systems. The whole mechanism exposes a lot of knobs to fine-tune the resource intensity of each controller.

**From the performance / runtime perspective:** for similar reasons I'm not sure that `virtctl` is easier to understand or use. You mentioned that `It is also costly from the users side, as they need to know about them and browse through potential a large number of them in the system` - but the user would need to browse a large number of options either way. Why do you think that browsing on `virtctl`, a custom binary, would be easier than browsing on CRDs, API that every Kubernetes user is familiar with?

To summarize: I agree that the "Kubernetes way" is not holy and we should always consider alternatives. On the other hand, I think that with any platform there are many automatic benefits gained by doing things in the "platform way". In other words, there should be a good enough reason not to go that way. In this specific case I don't see a huge value from not exposing a CRD.

Comment (davidvossel) [reply design-proposals/VM-cloning.md:46]: > AFAIK, we already have several imperative API commands (e.g. hotplug for storage, and soon for network) which do not use CRDs

hotplug is declarative.  it uses the VM/VMI status to make a volume request which models the declarative workflow for us.

> I am trying to understand if there are other alternatives to extend the functionality of kubevirt without adding CRD/s.

It you want to model a declarative workflow, using a new or existing CRD is the way to do it in k8s if you're a dev on the project itself. Something like a tekton pipeline would be the way to model it as a devops engineer. 

I'd prefer we keep the clone workflow outside of the VM object itself, similar to how we kept the snapshot/restore apis separate from the VM object. There's existing precedence within our ecosystem to support the pattern of using a CRD to model the clone workflow.

File: design-proposals/VM-cloning.md
Comment (davidvossel) [inline design-proposals/VM-cloning.md:152]: a `virtctl clone` client command might fall under enhancements too.

**Review by davidvossel — APPROVED** [review]
/lgtm

**Review by davidvossel — COMMENTED** [review]
/approve
/hold

@mhenriks can you give a `/lgtm` and a `/hold cancel` to this if you feel comfortable. We need 2 acks for designs.

File: design-proposals/VM-cloning.md
Comment (iholder-redhat) [reply design-proposals/VM-cloning.md]: Added. Thanks!

File: design-proposals/VM-cloning.md
Comment (jean-edouard) [inline design-proposals/VM-cloning.md:108]: The use of `VM'` for the clone name is confusing, even more below when reading `[...] VM's [...]`.
Also, in this line, there's a backtick instead of a `'`.

**Comment by davidvossel** [comment]
/approve
/hold

remove hold after fixing typos, great work!

File: design-proposals/VM-cloning.md
Comment (iholder-redhat) [reply design-proposals/VM-cloning.md:2]: Thanks!

File: design-proposals/VM-cloning.md
Comment (iholder-redhat) [reply design-proposals/VM-cloning.md:108]: Thanks. I'll use VM1 and VM2 instead

**Comment by iholder-redhat** [comment]
Bringing back lgtm after typo fixes

/lgtm

**Review by jean-edouard — APPROVED** [review]

File: design-proposals/VM-cloning.md
Comment (fabiand) [inline design-proposals/VM-cloning.md:50]: @iholder-redhat there are (at least) a few more things to consider:

- smbios serial numbers
- disk serial numbers
- vtpm (once it lands)
- disks(!) - what should happen to each during cloning?

In general: Cloning is effectively the process to take a VM with an identity, and regenerating (or stripping) all identity related parts from the API, at the same time, a snapshot of the disk needs to be taken at the time of cloning.

I'm missing the discussion around these elements in this design. Was this discussed elsewhere?

Comment (yaacov) [reply design-proposals/VM-cloning.md:50]: > what's being used in UI (cloning is fully implemented there [1]).

The UI is * very * naive, it create a dataVolume to copy the vms PVCs, and then attach the new PVC's.

Comment (iholder-redhat) [reply design-proposals/VM-cloning.md:50]: Thanks @yaacov.
What about the re-generation of things like smbios serial, disk serial, ssh hostname etc? How exactly is it handled from the UI perspective? Are you familiar with the QEMU function I was speaking of?

@awels - appreciate if you can take a look

Comment (yaacov) [reply design-proposals/VM-cloning.md:50]: > How exactly is it handled from the UI perspective?

It is not, the UI just create a copy.

>  Are you familiar with the QEMU function I was speaking of?

No, the UI can only interact with the kubevirt API, it does not interact with the launcher pods

Comment (iholder-redhat) [reply design-proposals/VM-cloning.md:50]: I think we need Alex's help here :)

In the meantime, pretty much all other logic (except for this magical QEMU function) is implemented here https://github.com/kubevirt/kubevirt/pull/7336. This PR is a draft and will stay that way until https://github.com/kubevirt/kubevirt/pull/7318 merges.

Feedback is much appreciated.

Comment (iholder-redhat) [reply design-proposals/VM-cloning.md:50]: > @iholder-redhat there are (at least) a few more things to consider:
> 
> * smbios serial numbers
> * disk serial numbers
> * vtpm (once it lands)
> * disks(!) - what should happen to each during cloning?
> 
> In general: Cloning is effectively the process to take a VM with an identity, and regenerating (or stripping) all identity related parts from the API, at the same time, a snapshot of the disk needs to be taken at the time of cloning.
> 
> I'm missing the discussion around these elements in this design. Was this discussed elsewhere?

Hey @fabiand,

I came back with some answers :smiley: 

> * smbios serial numbers

We can specify these in VMI's spec, but don't have to as the field is optional. If not set, libvirt will regenerate one for us. So all that's needed here is to delete the smbios serial from the new VM's spec.

> * disk serial numbers

Disk serial being the same for different VMs shouldn't be a problem, as I understood from @vladikr. They are only used as an identifier from within the VM, not from the outer world.

> * vtpm (once it lands)

Two options that come into mind:
1. To not migrate VMIs with vTPM at all (can maybe be a first step).
2. Have a condition in VMI that's indicating of vTPM use. To migrate such VMs the user would need to explicitly enable a feature gate or an option in KubevirtCR's spec.

Something along those lines. Maybe @jean-edouard can give us his thoughts here.

> * disks(!) - what should happen to each during cloning?

This is completely covered by the snapshot mechanism. In short: whenever we can we use CDI's smart-cloning, and in other situation we rely on volume snapshots.

Another thing to take care of is the MAC address that can simply be deleted if kube-mac-pool is available. Otherwise, there's an option in Clone's spec to provide a manual MAC address.

Comment (iholder-redhat) [reply design-proposals/VM-cloning.md:50]: Thanks @jean-edouard!
Indeed sounds like something we should tackle in the future :)

**Pull request merged**
